File: src/lib/deviceCatalog.ts
import catalogJson from '../assets/devices/catalog.json';

export type DeviceKind = 'watch' | 'strap' | 'ring' | 'band' | 'earbuds' | 'scale' | 'unknown';
export type DeviceCatalogStatus = 'active' | 'historical';
export type DeviceMatchStatus = 'exact' | 'alias' | 'unknown';

export interface DeviceCatalogEntry {
  catalog_id: string;
  canonical_name: string;
  display_name: string;
  name_zh?: string | null;
  kind: DeviceKind;
  model_codes: string[];
  aliases: string[];
  region: string[];
  status: DeviceCatalogStatus;
  supported: boolean;
  canonical_device_key?: string;
  official_page?: string;
  official_url: string;
  image_source_url?: string | null;
  asset_source?: string | null;
  image_key?: string | null;
  asset_hash?: string | null;
  checked_at: string;
  provenance: string;
}

export interface DeviceCatalogDocument {
  version: number;
  checked_at: string;
  sources: string[];
  devices: DeviceCatalogEntry[];
}

export interface DeviceCatalogMatch {
  entry: DeviceCatalogEntry;
  status: Exclude<DeviceMatchStatus, 'unknown'>;
  matched_by: 'model_code' | 'alias';
  matched_value: string;
}

export interface DeviceCatalogMatchInput {
  modelCodes?: string[];
  productNames?: string[];
  deviceNames?: string[];
  displayName?: string;
}

const document = catalogJson as DeviceCatalogDocument;

/** Versioned snapshot of the official catalog. No runtime network lookup is performed. */
export const deviceCatalog: readonly DeviceCatalogEntry[] = document.devices;
export const deviceCatalogVersion = document.version;
export const deviceCatalogCheckedAt = document.checked_at;
export const deviceCatalogSources: readonly string[] = document.sources;

/**
 * Assets are discovered at build time. Adding a catalog row only requires an
 * image pair in this directory; there is no 48-item hand-maintained import
 * list to drift out of sync.
 */
const imageModules = import.meta.glob('../assets/devices/*.webp', {
  eager: true,
  import: 'default',
  query: '?url',
}) as Record<string, string>;
const thumbnailModules = import.meta.glob('../assets/devices/*-thumb.png', {
  eager: true,
  import: 'default',
  query: '?url',
}) as Record<string, string>;

const keyFromPath = (path: string, suffix: string): string | null => {
  const normalizedPath = path.replace(/\\/g, '/');
  const match = normalizedPath.match(new RegExp(`/([^/]+)${suffix}$`));
  return match?.[1] ?? null;
};

/**
 * Vite emits imported assets according to the configured base. Keep a
 * defensive relative form for old/dev bundles that still contain a leading
 * slash; Tauri's asset protocol does not have the host-root path those URLs
 * imply. Data URLs (the original silhouette fallback) are left untouched.
 */
const runtimeAssetUrl = (source: string): string => {
  if (/^(?:data:|https?:|asset:|blob:)/u.test(source)) return source;
  return source.replace(/^\/+/, './');
};

export const localDeviceAssets: Readonly<Record<string, string>> = Object.freeze(
  Object.fromEntries(
    Object.entries(imageModules)
      .map(([path, source]) => [keyFromPath(path, '\\.webp'), runtimeAssetUrl(source)] as const)
      .filter((entry): entry is readonly [string, string] => Boolean(entry[0])),
  ),
);

export const localDeviceThumbnails: Readonly<Record<string, string>> = Object.freeze(
  Object.fromEntries(
    Object.entries(thumbnailModules)
      .map(([path, source]) => [keyFromPath(path, '-thumb\\.png'), runtimeAssetUrl(source)] as const)
      .filter((entry): entry is readonly [string, string] => Boolean(entry[0])),
  ),
);

export const normalizeDeviceText = (value: string): string =>
  value.normalize('NFKC').toLocaleLowerCase().replace(/[\u0000-\u001f]/g, '').replace(/[^\p{L}\p{N}]+/gu, '');

const deviceWords = (value: string): string[] =>
  value.normalize('NFKC').toLocaleLowerCase().match(/[\p{L}\p{N}]+/gu) ?? [];

const containsCompleteAlias = (displayName: string, alias: string): boolean => {
  const aliasWords = deviceWords(alias);
  // A single generic word such as "Balance" is intentionally not enough to
  // infer a product from a nickname. Numbered/two-word aliases are stable.
  if (aliasWords.length < 2 && !aliasWords.some((word) => /\d/u.test(word))) return false;
  // User nicknames may put a CJK prefix directly before the Latin product
  // name (for example, "凌苍的T-Rex 3").  The token matcher merges that
  // prefix with the first Latin word, so also scan the punctuation-free form
  // while retaining ASCII boundaries to reject "T-Rex 30".
  const display = normalizeDeviceText(displayName);
  const needle = normalizeDeviceText(alias);
  if (!needle) return false;
  let offset = 0;
  while (offset < display.length) {
    const found = display.indexOf(needle, offset);
    if (found < 0) break;
    const start = found;
    const end = start + needle.length;
    const before = start > 0 ? display[start - 1] : undefined;
    const after = end < display.length ? display[end] : undefined;
    const asciiBoundary = (value: string | undefined) => value === undefined || !/[A-Za-z0-9]/u.test(value);
    if (asciiBoundary(before) && asciiBoundary(after)) return true;
    offset = end;
  }
  return false;
};

const unique = (values: readonly (string | undefined | null)[]): string[] =>
  values.map((value) => value?.trim()).filter((value): value is string => Boolean(value));

/**
 * Match in the same strict order used by the Rust IPC parser:
 * stable model code, exact product/device alias, then a complete alias in a
 * display nickname. Generic fuzzy matching is deliberately not used.
 */
export function matchDeviceCatalog(input: DeviceCatalogMatchInput): DeviceCatalogMatch | null {
  const matchable = (item: DeviceCatalogEntry): boolean => item.supported && item.status === 'active';
  const modelCodes = unique(input.modelCodes ?? []);
  for (const candidate of modelCodes) {
    const normalized = normalizeDeviceText(candidate);
    if (!normalized) continue;
    const entry = deviceCatalog.find((item) =>
      matchable(item) && item.model_codes.some((code) => normalizeDeviceText(code) === normalized),
    );
    if (entry) {
      return { entry, status: 'exact', matched_by: 'model_code', matched_value: candidate };
    }
  }

  const exactNames = unique([...(input.productNames ?? []), ...(input.deviceNames ?? [])]);
  for (const candidate of exactNames) {
    const normalized = normalizeDeviceText(candidate);
    if (!normalized) continue;
    const entry = deviceCatalog.find((item) =>
      matchable(item)
      && [item.display_name, item.name_zh, ...item.aliases]
        .filter((alias): alias is string => Boolean(alias))
        .some((alias) => normalizeDeviceText(alias) === normalized),
    );
    if (entry) {
      return { entry, status: 'alias', matched_by: 'alias', matched_value: candidate };
    }
  }

  if (input.displayName) {
    const entry = deviceCatalog.find((item) =>
      matchable(item)
      && [item.display_name, item.name_zh, ...item.aliases]
        .filter((alias): alias is string => Boolean(alias))
        .some((alias) => containsCompleteAlias(input.displayName!, alias)),
    );
    if (entry) {
      return { entry, status: 'alias', matched_by: 'alias', matched_value: input.displayName };
    }
  }
  return null;
}

export function deviceImageFor(kind: DeviceKind | string | undefined, imageKey?: string | null): string {
  if (imageKey && localDeviceAssets[imageKey]) return localDeviceAssets[imageKey];
  return deviceFallbackFor(kind);
}

/** Missing images are rendered by DeviceVisual's inline, CSP-safe SVG fallback. */
export function deviceFallbackFor(_kind: DeviceKind | string | undefined): string {
  return '';
}

export function deviceThumbnailFor(kind: DeviceKind | string | undefined, imageKey?: string | null): string {
  if (imageKey && localDeviceThumbnails[imageKey]) return localDeviceThumbnails[imageKey];
  return deviceImageFor(kind, imageKey);
}

File: src-tauri/src/device_catalog.rs
use serde::Deserialize;
use std::sync::OnceLock;

const CATALOG_JSON: &str = include_str!("../../src/assets/devices/catalog.json");

#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize)]
pub struct CatalogDocument {
    pub version: u32,
    pub checked_at: String,
    pub sources: Vec<String>,
    pub devices: Vec<CatalogEntry>,
}

#[allow(dead_code)]
#[derive(Debug, Clone, Deserialize)]
pub struct CatalogEntry {
    pub catalog_id: String,
    pub canonical_name: String,
    pub display_name: String,
    #[serde(default)]
    pub name_zh: Option<String>,
    pub kind: String,
    #[serde(default)]
    pub model_codes: Vec<String>,
    pub aliases: Vec<String>,
    pub region: Vec<String>,
    pub status: String,
    pub supported: bool,
    #[serde(default)]
    pub canonical_device_key: Option<String>,
    #[serde(default)]
    pub official_page: Option<String>,
    pub official_url: String,
    #[serde(default)]
    pub image_source_url: Option<String>,
    #[serde(default)]
    pub asset_source: Option<String>,
    #[serde(default)]
    pub image_key: Option<String>,
    #[serde(default)]
    pub asset_hash: Option<String>,
    pub checked_at: String,
    pub provenance: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CatalogMatchStatus {
    Exact,
    Alias,
}

#[derive(Debug, Clone)]
pub struct CatalogMatch<'a> {
    pub entry: &'a CatalogEntry,
    pub status: CatalogMatchStatus,
}

#[derive(Debug, Default)]
pub struct CatalogMatchInput<'a> {
    pub model_codes: Vec<&'a str>,
    pub product_names: Vec<&'a str>,
    pub device_names: Vec<&'a str>,
    pub display_name: Option<&'a str>,
}

fn document() -> &'static CatalogDocument {
    static DOCUMENT: OnceLock<CatalogDocument> = OnceLock::new();
    DOCUMENT.get_or_init(|| {
        serde_json::from_str(CATALOG_JSON).expect("bundled device catalog must be valid JSON")
    })
}

pub fn catalog_entries() -> &'static [CatalogEntry] {
    &document().devices
}

pub fn normalize_model(value: &str) -> String {
    value
        .chars()
        .flat_map(char::to_lowercase)
        .filter(|character| character.is_alphanumeric())
        .collect()
}

fn words(value: &str) -> Vec<String> {
    value
        .split(|character: char| !character.is_alphanumeric())
        .filter(|word| !word.is_empty())
        .map(|word| word.to_lowercase())
        .collect()
}

fn contains_complete_alias(display_name: &str, alias: &str) -> bool {
    let alias_words = words(alias);
    // A single generic word (for example, "Balance") cannot identify one
    // product from a nickname. Numbered or multi-word aliases are stable.
    if alias_words.len() < 2
        && !alias_words
            .iter()
            .any(|word| word.chars().any(|c| c.is_ascii_digit()))
    {
        return false;
    }
    // Product aliases can be embedded in a user nickname written in CJK
    // characters (for example, "凌苍的T-Rex 3").  The old word-window
    // matcher merged the CJK prefix and the first Latin token, so it never
    // saw the complete alias.  Match the punctuation-free alias while still
    // requiring ASCII boundaries, which rejects near misses such as
    // "T-Rex 30" without inventing a product for arbitrary text.
    let display = normalize_model(display_name);
    let needle = normalize_model(alias);
    if needle.is_empty() {
        return false;
    }
    let mut offset = 0;
    while let Some(found) = display[offset..].find(&needle) {
        let start = offset + found;
        let end = start + needle.len();
        let before = display[..start].chars().next_back();
        let after = display[end..].chars().next();
        let ascii_boundary = |character: Option<char>| {
            character
                .map(|value| !value.is_ascii_alphanumeric())
                .unwrap_or(true)
        };
        if ascii_boundary(before) && ascii_boundary(after) {
            return true;
        }
        offset = start + needle.len();
        if offset >= display.len() {
            break;
        }
    }
    false
}

pub fn match_catalog(input: &CatalogMatchInput<'_>) -> Option<CatalogMatch<'static>> {
    for candidate in &input.model_codes {
        let normalized = normalize_model(candidate);
        if normalized.is_empty() {
            continue;
        }
        if let Some(entry) = catalog_entries().iter().find(|entry| {
            entry.supported
                && entry.status == "active"
                && entry
                    .model_codes
                    .iter()
                    .any(|code| normalize_model(code) == normalized)
        }) {
            return Some(CatalogMatch {
                entry,
                status: CatalogMatchStatus::Exact,
            });
        }
    }

    for candidate in input.product_names.iter().chain(input.device_names.iter()) {
        let normalized = normalize_model(candidate);
        if normalized.is_empty() {
            continue;
        }
        if let Some(entry) = catalog_entries().iter().find(|entry| {
            entry.supported
                && entry.status == "active"
                && std::iter::once(&entry.display_name)
                    .chain(entry.aliases.iter())
                    .chain(entry.name_zh.iter())
                    .any(|alias| normalize_model(alias) == normalized)
        }) {
            return Some(CatalogMatch {
                entry,
                status: CatalogMatchStatus::Alias,
            });
        }
    }

    let display_name = input.display_name?;
    catalog_entries()
        .iter()
        .filter(|entry| entry.supported && entry.status == "active")
        .flat_map(|entry| {
            std::iter::once(&entry.display_name)
                .chain(entry.aliases.iter())
                .chain(entry.name_zh.iter())
                .filter_map(move |alias| {
                    contains_complete_alias(display_name, alias)
                        .then_some((alias.split_whitespace().count(), entry))
                })
        })
        .max_by_key(|(length, _)| *length)
        .map(|(_, entry)| CatalogMatch {
            entry,
            status: CatalogMatchStatus::Alias,
        })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn catalog_contains_real_devices_and_version() {
        assert!(document().version >= 1);
        assert!(document().checked_at.starts_with("2026-"));
        assert!(catalog_entries()
            .iter()
            .any(|entry| entry.catalog_id == "amazfit-t-rex-3"));
        assert!(catalog_entries()
            .iter()
            .any(|entry| entry.catalog_id == "amazfit-helio-strap"));
        assert!(catalog_entries()
            .iter()
            .any(|entry| entry.catalog_id == "amazfit-helio-ring"));
    }

    #[test]
    fn matching_uses_code_then_exact_names_then_complete_display_alias() {
        let exact = match_catalog(&CatalogMatchInput {
            model_codes: vec!["A2323"],
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(exact.status, CatalogMatchStatus::Exact);
        assert_eq!(exact.entry.catalog_id, "amazfit-t-rex-3");

        let alias = match_catalog(&CatalogMatchInput {
            product_names: vec!["Helio Strap"],
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(alias.status, CatalogMatchStatus::Alias);
        assert_eq!(alias.entry.catalog_id, "amazfit-helio-strap");

        let display = match_catalog(&CatalogMatchInput {
            display_name: Some("我的 T-Rex 3"),
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(display.entry.catalog_id, "amazfit-t-rex-3");

        let cjk_t_rex = match_catalog(&CatalogMatchInput {
            display_name: Some("凌苍的T-Rex 3"),
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(cjk_t_rex.entry.catalog_id, "amazfit-t-rex-3");

        let cjk_helio = match_catalog(&CatalogMatchInput {
            display_name: Some("凌苍的Helio Strap"),
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(cjk_helio.entry.catalog_id, "amazfit-helio-strap");

        let pro = match_catalog(&CatalogMatchInput {
            product_names: vec!["T-Rex 3 Pro 48mm"],
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(pro.entry.catalog_id, "amazfit-t-rex-3-pro-48-44mm");

        let ultra = match_catalog(&CatalogMatchInput {
            display_name: Some("我的 Amazfit T-Rex Ultra"),
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(ultra.entry.catalog_id, "amazfit-t-rex-ultra-47mm");

        let square = match_catalog(&CatalogMatchInput {
            product_names: vec!["Active 2 Square"],
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(square.entry.catalog_id, "amazfit-active-2-square");

        let bip_pro = match_catalog(&CatalogMatchInput {
            product_names: vec!["Bip 3 Pro"],
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(bip_pro.entry.catalog_id, "amazfit-bip-3-pro");

        let bip_five = match_catalog(&CatalogMatchInput {
            display_name: Some("我的 Amazfit Bip 5"),
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(bip_five.entry.catalog_id, "amazfit-bip-5-46mm");

        let bip_six = match_catalog(&CatalogMatchInput {
            display_name: Some("我的 Bip 6"),
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(bip_six.entry.catalog_id, "amazfit-bip-6");

        let black = match_catalog(&CatalogMatchInput {
            product_names: vec!["GTR 4 46mm Black"],
            ..CatalogMatchInput::default()
        })
        .unwrap();
        assert_eq!(black.entry.catalog_id, "amazfit-gtr-4-46mm-black");

        assert!(match_catalog(&CatalogMatchInput {
            display_name: Some("未知手环"),
            ..CatalogMatchInput::default()
        })
        .is_none());

        assert!(match_catalog(&CatalogMatchInput {
            product_names: vec!["Helio Armband"],
            ..CatalogMatchInput::default()
        })
        .is_none());
    }
}
